File: services/routing_service.py
import math
from urllib.parse import quote

import requests


OSRM_BASE_URL = "https://router.project-osrm.org"
OSRM_ROUTE_MODE = "driving"
OSRM_TIMEOUT_SECONDS = 12


def build_cache_key(stop_a: dict, stop_b: dict) -> str:
    return f"{stop_a.get('id', '')}->{stop_b.get('id', '')}"


def get_travel_segment(
    stop_a: dict,
    stop_b: dict,
    cache: dict | None = None,
    route_mode: str = "car",
    timeout_seconds: int = OSRM_TIMEOUT_SECONDS,
):
    """Lädt die Fahrzeit für ein Segment über OSRM und nutzt dabei einen einfachen Cache.

    Der Cache wird pro Segment unter `stopA_id->stopB_id` abgelegt. Basis-URL und Timeout
    sind als Konstanten gekapselt, damit ein späterer Wechsel des Routers zentral möglich ist.
    """
    cache = cache if isinstance(cache, dict) else {}
    cache_key = build_cache_key(stop_a, stop_b)
    cached_minutes = cache.get(cache_key)
    if cached_minutes is not None:
        try:
            return {
                "cache_key": cache_key,
                "minutes": int(round(float(cached_minutes))),
                "distance_km": None,
                "cached": True,
                "error": "",
            }
        except Exception:
            pass

    lat_a = stop_a.get("lat")
    lon_a = stop_a.get("lon", stop_a.get("lng"))
    lat_b = stop_b.get("lat")
    lon_b = stop_b.get("lon", stop_b.get("lng"))
    if None in (lat_a, lon_a, lat_b, lon_b):
        return {"cache_key": cache_key, "minutes": None, "distance_km": None, "cached": False, "error": "Koordinaten fehlen"}

    profile = OSRM_ROUTE_MODE if route_mode == "car" else route_mode
    coords = f"{float(lon_a)},{float(lat_a)};{float(lon_b)},{float(lat_b)}"
    url = f"{OSRM_BASE_URL}/route/v1/{quote(profile)}/{coords}?overview=false"

    try:
        response = requests.get(url, timeout=timeout_seconds)
        response.raise_for_status()
        data = response.json()
        route = (data.get("routes") or [{}])[0]
        duration_seconds = route.get("duration")
        distance_meters = route.get("distance")
        if duration_seconds is None:
            raise ValueError("Keine Dauer in Routing-Antwort")
        minutes = max(1, int(round(float(duration_seconds) / 60.0)))
        distance_km = None
        if distance_meters is not None:
            distance_km = round(float(distance_meters) / 1000.0, 1)
        cache[cache_key] = minutes
        return {
            "cache_key": cache_key,
            "minutes": minutes,
            "distance_km": distance_km,
            "cached": False,
            "error": "",
        }
    except Exception as exc:
        return {
            "cache_key": cache_key,
            "minutes": None,
            "distance_km": None,
            "cached": False,
            "error": str(exc),
        }


def estimate_distance_km(stop_a: dict, stop_b: dict):
    lat_a = stop_a.get("lat")
    lon_a = stop_a.get("lon", stop_a.get("lng"))
    lat_b = stop_b.get("lat")
    lon_b = stop_b.get("lon", stop_b.get("lng"))
    if None in (lat_a, lon_a, lat_b, lon_b):
        return None

    radius_km = 6371.0
    d_lat = math.radians(float(lat_b) - float(lat_a))
    d_lon = math.radians(float(lon_b) - float(lon_a))
    a = (
        math.sin(d_lat / 2) ** 2
        + math.cos(math.radians(float(lat_a)))
        * math.cos(math.radians(float(lat_b)))
        * math.sin(d_lon / 2) ** 2
    )
    c = 2 * math.atan2(math.sqrt(a), math.sqrt(1 - a))
    return round(radius_km * c, 1)

File: GAWELA-Tourenplaner-2.0/src/GAWELA.Tourenplaner2/Services/RoutingService.cs
using System.Net.Http.Json;
using System.Text.Json.Serialization;

namespace GAWELA.Tourenplaner2.Services;

public sealed class TravelSegment
{
    public string CacheKey { get; set; } = string.Empty;
    public int? Minutes { get; set; }
    public double? DistanceKm { get; set; }
    public bool Cached { get; set; }
    public string Error { get; set; } = string.Empty;
}

public static class RoutingService
{
    private static readonly HttpClient HttpClient = new() { Timeout = TimeSpan.FromSeconds(12) };
    private const string BaseUrl = "https://router.project-osrm.org";

    public static string BuildCacheKey(string stopAId, string stopBId) => $"{stopAId}->{stopBId}";

    public static async Task<TravelSegment> GetTravelSegmentAsync(
        string stopAId,
        double? latA,
        double? lonA,
        string stopBId,
        double? latB,
        double? lonB,
        Dictionary<string, int> cache,
        string routeMode = "car")
    {
        var cacheKey = BuildCacheKey(stopAId, stopBId);
        if (cache.TryGetValue(cacheKey, out var cachedMinutes))
        {
            return new TravelSegment { CacheKey = cacheKey, Minutes = cachedMinutes, Cached = true };
        }

        if (!latA.HasValue || !lonA.HasValue || !latB.HasValue || !lonB.HasValue)
        {
            return new TravelSegment { CacheKey = cacheKey, Error = "Koordinaten fehlen" };
        }

        try
        {
            var profile = routeMode == "car" ? "driving" : routeMode;
            var coords = $"{lonA.Value},{latA.Value};{lonB.Value},{latB.Value}";
            var url = $"{BaseUrl}/route/v1/{Uri.EscapeDataString(profile)}/{coords}?overview=false";
            var response = await HttpClient.GetFromJsonAsync<OsrmResponse>(url);

            var route = response?.Routes?.FirstOrDefault();
            if (route?.Duration is null)
            {
                throw new InvalidOperationException("Keine Dauer in Routing-Antwort");
            }

            var minutes = Math.Max(1, (int)Math.Round(route.Duration.Value / 60.0));
            var distanceKm = route.Distance.HasValue ? Math.Round(route.Distance.Value / 1000.0, 1) : null;
            cache[cacheKey] = minutes;

            return new TravelSegment { CacheKey = cacheKey, Minutes = minutes, DistanceKm = distanceKm, Cached = false };
        }
        catch (Exception ex)
        {
            return new TravelSegment { CacheKey = cacheKey, Error = ex.Message, Cached = false };
        }
    }

    public static double? EstimateDistanceKm(double? latA, double? lonA, double? latB, double? lonB)
    {
        if (!latA.HasValue || !lonA.HasValue || !latB.HasValue || !lonB.HasValue)
        {
            return null;
        }

        const double radiusKm = 6371.0;
        var dLat = DegreesToRadians(latB.Value - latA.Value);
        var dLon = DegreesToRadians(lonB.Value - lonA.Value);

        var a = Math.Sin(dLat / 2) * Math.Sin(dLat / 2) +
                Math.Cos(DegreesToRadians(latA.Value)) * Math.Cos(DegreesToRadians(latB.Value)) *
                Math.Sin(dLon / 2) * Math.Sin(dLon / 2);

        var c = 2 * Math.Atan2(Math.Sqrt(a), Math.Sqrt(1 - a));
        return Math.Round(radiusKm * c, 1);
    }

    private static double DegreesToRadians(double degrees) => degrees * Math.PI / 180.0;

    private sealed class OsrmResponse
    {
        [JsonPropertyName("routes")]
        public List<OsrmRoute>? Routes { get; set; }
    }

    private sealed class OsrmRoute
    {
        [JsonPropertyName("duration")]
        public double? Duration { get; set; }

        [JsonPropertyName("distance")]
        public double? Distance { get; set; }
    }
}
